pkg/grpc/proxy/handler.go

```diff
@@ -186,26 +199,27 @@ func (s *handler) forwardClientToServer(src grpc.ClientStream, dst grpc.ServerSt
 	return ret
 }
 
-func (s *handler) forwardServerToClient(src grpc.ServerStream, dst grpc.ClientStream) chan error {
+func (s *handler) forwardServerToClient(src grpc.ServerStream, dst grpc.ClientStream, requestID string) chan error {
 	ret := make(chan error, 1)
 	go func() {
 		f := &codec.Frame{}
-		bufferedSends := s.bufferedCalls
-		if len(bufferedSends) > 0 {
-			for _, msg := range s.bufferedCalls {
-				if err := dst.SendMsg(msg); err != nil {
-					ret <- err
-					return
-				}
+		syncMapValue, _ := s.bufferedCalls.Load(requestID)
+		bufferedFrames := syncMapValue.([]interface{})
+		for _, msg := range bufferedFrames {
+			if err := dst.SendMsg(msg); err != nil {
+				ret <- err
+				return
 			}
 		}
 		for i := 0; ; i++ {
 			if err := src.RecvMsg(f); err != nil {
+				s.bufferedCalls.Store(requestID, bufferedFrames)
 				ret <- err // this can be io.EOF which is happy case
 				break
 			}
-			s.bufferedCalls = append(s.bufferedCalls, f)
+			bufferedFrames = append(bufferedFrames, f)
```

Comment:
Is there a concurrency issue here if >1 goroutine is accessing the underlying buffer?

Reply:
It's a local variable in the goroutine which is referenced out of the sync map by a uuid. For two goroutines to access the same underlying buffer, we'd need to kick off two streams from the same request at the same time, which I don't believe is possible. 

https://github.com/dapr/dapr/pull/4829/files#diff-b4595741c39d4944ac576f1228c698c0871002bfb2d2bbf6eab939aa66dac7feR207

Reply:
There should likely be a lock here per request ID

Reply:
> It's a local variable in the goroutine which is referenced out of the sync map by a uuid. For two goroutines to access the same underlying buffer, we'd need to kick off two streams from the same request at the same time, which I don't believe is possible.
> 
> https://github.com/dapr/dapr/pull/4829/files#diff-b4595741c39d4944ac576f1228c698c0871002bfb2d2bbf6eab939aa66dac7feR207

That assumption should hold true.

Reply:
@yaron2 - If we have multiple goroutines from the same request in here, wouldn't we also be sending/receiving incorrectly because we'd have multiple listeners on the same stream? Or is this a "just for safety's sake" thing?